In [25]:
import pandas as pd
import numpy as np 
import os 

In [26]:
# 원본 데이터 불러오기
# 파일경로명 다르게 하시면 실행됩니다!

chicago = pd.read_csv('../input/preprocessing_US_Chicago_Traffic_Crashes.csv')

In [27]:
chicago

POSTED_SPEED_LIMIT TRAFFIC_CONTROL_DEVICE      DEVICE_CONDITION  \
0                       10            NO CONTROLS           NO CONTROLS   
1                       45            NO CONTROLS           NO CONTROLS   
2                       35      STOP SIGN/FLASHER  FUNCTIONING PROPERLY   
3                       30         TRAFFIC SIGNAL  FUNCTIONING PROPERLY   
4                       35            NO CONTROLS           NO CONTROLS   
...                    ...                    ...                   ...   
433054                  30            NO CONTROLS           NO CONTROLS   
433055                  30            NO CONTROLS           NO CONTROLS   
433056                  30         TRAFFIC SIGNAL  FUNCTIONING PROPERLY   
433057                  30            NO CONTROLS           NO CONTROLS   
433058                  15            NO CONTROLS           NO CONTROLS   

       WEATHER_CONDITION      LIGHTING_CONDITION  LANE_CNT  \
0                  CLEAR                DARKNESS       NaN   
1                  CLEAR                DAYLIGHT       NaN   
2                  CLEAR                DAYLIGHT       4.0   
3                  CLEAR  DARKNESS, LIGHTED ROAD       4.0   
4                  CLEAR                DAYLIGHT       NaN   
...                  ...                     ...       ...   
433054             CLEAR                DAYLIGHT       NaN   
433055             CLEAR                DAYLIGHT       NaN   
433056             CLEAR                DAYLIGHT       NaN   
433057             CLEAR                DAYLIGHT       NaN   
433058             CLEAR                DAYLIGHT       NaN   

       ROADWAY_SURFACE_COND ROAD_DEFECT  CRASH_HOUR  
0                       DRY  NO DEFECTS          18  
1                       DRY  NO DEFECTS          16  
2                       DRY  NO DEFECTS          16  
3                       DRY  NO DEFECTS          22  
4                       DRY  NO DEFECTS          14  
...                     ...         ...         ...  
433054                  DRY  NO DEFECTS          16  
433055                  DRY  NO DEFECTS          14  
433056                  DRY  NO DEFECTS          10  
433057                  DRY  NO DEFECTS          18  
433058                  DRY  NO DEFECTS           8  

[433059 rows x 9 columns]

In [28]:
chicago.isnull().sum()

# 결측치 제거1
# uk_accidents_2018 = df.drop(["Accident_Index", "Location_Easting_OSGR", "Location_Northing_OSGR", "Police_Force", "Did_Police_Officer_Attend_Scene_of_Accident", "LSOA_of_Accident_Location"], axis=1)

# 결측치 제거2
# 경도, 위도 결측치 제거
# df = df.dropna(subset = ["Longitude", "Latitude"])

POSTED_SPEED_LIMIT             0
TRAFFIC_CONTROL_DEVICE         0
DEVICE_CONDITION               0
WEATHER_CONDITION              0
LIGHTING_CONDITION             0
LANE_CNT                  234102
ROADWAY_SURFACE_COND           0
ROAD_DEFECT                    0
CRASH_HOUR                     0
dtype: int64

In [29]:
chicago.dtypes

POSTED_SPEED_LIMIT          int64
TRAFFIC_CONTROL_DEVICE     object
DEVICE_CONDITION           object
WEATHER_CONDITION          object
LIGHTING_CONDITION         object
LANE_CNT                  float64
ROADWAY_SURFACE_COND       object
ROAD_DEFECT                object
CRASH_HOUR                  int64
dtype: object

<br>
# 1. POSTED_SPEED_LIMIT

**상의 ㄱㄱ**

In [48]:
chicago["POSTED_SPEED_LIMIT"].value_counts(normalize=True) * 100

30    73.701736
35     6.873890
25     6.017887
20     3.857211
15     3.449184
10     2.039214
0      1.538589
40     0.897568
5      0.807973
45     0.593453
55     0.093752
3      0.024477
50     0.020551
9      0.020551
99     0.015240
39     0.010622
1      0.008082
60     0.006004
2      0.004387
32     0.003002
65     0.002540
24     0.002078
33     0.002078
6      0.001616
11     0.001155
34     0.001155
36     0.000924
70     0.000693
18     0.000462
14     0.000462
12     0.000462
7      0.000462
31     0.000462
4      0.000231
49     0.000231
63     0.000231
16     0.000231
23     0.000231
26     0.000231
29     0.000231
38     0.000231
22     0.000231
Name: POSTED_SPEED_LIMIT, dtype: float64

<br>
# 2. TRAFFIC_CONTROL_DEVICE

In [30]:
# DataFrame 복사
transform = chicago

def change2(x):
    if x=="NO CONTROLS": return 1
    elif x=="TRAFFIC SIGNAL": return 2
    elif x=="STOP SIGN/FLASHER": return 3
    elif x=="OTHER": return 4
    elif x=="LANE USE MARKING": return 5
    elif x=="YIELD": return 6
    elif x=="OTHER REG. SIGN": return 7
    elif x=="OTHER WARNING SIGN": return 8
    elif x=="RAILROAD CROSSING GATE": return 9
    elif x=="POLICE/FLAGMAN": return 10
    elif x=="SCHOOL ZONE": return 11
    elif x=="PEDESTRIAN CROSSING SIGN": return 12
    elif x=="DELINEATORS": return 13
    elif x=="OTHER RAILROAD CROSSING": return 14
    elif x=="FLASHING CONTROL SIGNAL": return 15
    elif x=="RR CROSSING SIGN": return 16
    elif x=="NO PASSING": return 17
    elif x=="BICYCLE CROSSING SIGN": return 18
    elif x=="UNKNOWN": return 0
    
transform["TRAFFIC_CONTROL_DEVICE"] = transform["TRAFFIC_CONTROL_DEVICE"].apply(change2)

In [31]:
transform

POSTED_SPEED_LIMIT  TRAFFIC_CONTROL_DEVICE      DEVICE_CONDITION  \
0                       10                       1           NO CONTROLS   
1                       45                       1           NO CONTROLS   
2                       35                       3  FUNCTIONING PROPERLY   
3                       30                       2  FUNCTIONING PROPERLY   
4                       35                       1           NO CONTROLS   
...                    ...                     ...                   ...   
433054                  30                       1           NO CONTROLS   
433055                  30                       1           NO CONTROLS   
433056                  30                       2  FUNCTIONING PROPERLY   
433057                  30                       1           NO CONTROLS   
433058                  15                       1           NO CONTROLS   

       WEATHER_CONDITION      LIGHTING_CONDITION  LANE_CNT  \
0                  CLEAR                DARKNESS       NaN   
1                  CLEAR                DAYLIGHT       NaN   
2                  CLEAR                DAYLIGHT       4.0   
3                  CLEAR  DARKNESS, LIGHTED ROAD       4.0   
4                  CLEAR                DAYLIGHT       NaN   
...                  ...                     ...       ...   
433054             CLEAR                DAYLIGHT       NaN   
433055             CLEAR                DAYLIGHT       NaN   
433056             CLEAR                DAYLIGHT       NaN   
433057             CLEAR                DAYLIGHT       NaN   
433058             CLEAR                DAYLIGHT       NaN   

       ROADWAY_SURFACE_COND ROAD_DEFECT  CRASH_HOUR  
0                       DRY  NO DEFECTS          18  
1                       DRY  NO DEFECTS          16  
2                       DRY  NO DEFECTS          16  
3                       DRY  NO DEFECTS          22  
4                       DRY  NO DEFECTS          14  
...                     ...         ...         ...  
433054                  DRY  NO DEFECTS          16  
433055                  DRY  NO DEFECTS          14  
433056                  DRY  NO DEFECTS          10  
433057                  DRY  NO DEFECTS          18  
433058                  DRY  NO DEFECTS           8  

[433059 rows x 9 columns]

In [32]:
transform["TRAFFIC_CONTROL_DEVICE"].value_counts(normalize=True) * 100

1     57.565597
2     27.853249
3      9.874174
0      3.220346
4      0.600842
5      0.283102
6      0.143860
7      0.098370
8      0.090519
9      0.068351
10     0.039025
11     0.037870
12     0.037639
13     0.027479
14     0.023092
15     0.022399
16     0.007389
17     0.004156
18     0.002540
Name: TRAFFIC_CONTROL_DEVICE, dtype: float64

<br>
# 3. DEVICE_CONDITION

In [33]:
def change3(x):
    if x=="NO CONTROLS": return 1
    elif x=="FUNCTIONING PROPERLY": return 2
    elif x=="OTHER": return 3
    elif x=="FUNCTIONING IMPROPERLY": return 4
    elif x=="NOT FUNCTIONING": return 5
    elif x=="WORN REFLECTIVE MATERIAL": return 6
    elif x=="MISSING": return 7
    elif x=="UNKNOWN": return 0
    
transform["DEVICE_CONDITION"] = transform["DEVICE_CONDITION"].apply(change3)

In [34]:
transform

POSTED_SPEED_LIMIT  TRAFFIC_CONTROL_DEVICE  DEVICE_CONDITION  \
0                       10                       1                 1   
1                       45                       1                 1   
2                       35                       3                 2   
3                       30                       2                 2   
4                       35                       1                 1   
...                    ...                     ...               ...   
433054                  30                       1                 1   
433055                  30                       1                 1   
433056                  30                       2                 2   
433057                  30                       1                 1   
433058                  15                       1                 1   

       WEATHER_CONDITION      LIGHTING_CONDITION  LANE_CNT  \
0                  CLEAR                DARKNESS       NaN   
1                  CLEAR                DAYLIGHT       NaN   
2                  CLEAR                DAYLIGHT       4.0   
3                  CLEAR  DARKNESS, LIGHTED ROAD       4.0   
4                  CLEAR                DAYLIGHT       NaN   
...                  ...                     ...       ...   
433054             CLEAR                DAYLIGHT       NaN   
433055             CLEAR                DAYLIGHT       NaN   
433056             CLEAR                DAYLIGHT       NaN   
433057             CLEAR                DAYLIGHT       NaN   
433058             CLEAR                DAYLIGHT       NaN   

       ROADWAY_SURFACE_COND ROAD_DEFECT  CRASH_HOUR  
0                       DRY  NO DEFECTS          18  
1                       DRY  NO DEFECTS          16  
2                       DRY  NO DEFECTS          16  
3                       DRY  NO DEFECTS          22  
4                       DRY  NO DEFECTS          14  
...                     ...         ...         ...  
433054                  DRY  NO DEFECTS          16  
433055                  DRY  NO DEFECTS          14  
433056                  DRY  NO DEFECTS          10  
433057                  DRY  NO DEFECTS          18  
433058                  DRY  NO DEFECTS           8  

[433059 rows x 9 columns]

In [35]:
transform["DEVICE_CONDITION"].value_counts(normalize=True) * 100

1    58.117716
2    34.711667
0     5.448680
3     0.774721
4     0.554659
5     0.334596
6     0.045259
7     0.012700
Name: DEVICE_CONDITION, dtype: float64

<br>

# 4. WEATHER_CONDITION

In [36]:
def change4(x):
    if x=="CLEAR": return 1
    elif x=="RAIN": return 2
    elif x=="SNOW": return 3
    elif x=="CLOUDY/OVERCAST": return 4
    elif x=="OTHER": return 5
    elif x=="FOG/SMOKE/HAZE": return 6
    elif x=="SLEET/HAIL": return 7
    elif x=="FREEZING RAIN/DRIZZLE": return 8
    elif x=="SEVERE CROSS WIND GATE": return 9
    elif x=="BLOWING SNOW": return 10
    elif x=="BLOWING SAND, SOIL, DIRT": return 11
    elif x=="UNKNOWN": return 0
    
transform["WEATHER_CONDITION"] = transform["WEATHER_CONDITION"].apply(change4)

In [37]:
transform

POSTED_SPEED_LIMIT  TRAFFIC_CONTROL_DEVICE  DEVICE_CONDITION  \
0                       10                       1                 1   
1                       45                       1                 1   
2                       35                       3                 2   
3                       30                       2                 2   
4                       35                       1                 1   
...                    ...                     ...               ...   
433054                  30                       1                 1   
433055                  30                       1                 1   
433056                  30                       2                 2   
433057                  30                       1                 1   
433058                  15                       1                 1   

        WEATHER_CONDITION      LIGHTING_CONDITION  LANE_CNT  \
0                       1                DARKNESS       NaN   
1                       1                DAYLIGHT       NaN   
2                       1                DAYLIGHT       4.0   
3                       1  DARKNESS, LIGHTED ROAD       4.0   
4                       1                DAYLIGHT       NaN   
...                   ...                     ...       ...   
433054                  1                DAYLIGHT       NaN   
433055                  1                DAYLIGHT       NaN   
433056                  1                DAYLIGHT       NaN   
433057                  1                DAYLIGHT       NaN   
433058                  1                DAYLIGHT       NaN   

       ROADWAY_SURFACE_COND ROAD_DEFECT  CRASH_HOUR  
0                       DRY  NO DEFECTS          18  
1                       DRY  NO DEFECTS          16  
2                       DRY  NO DEFECTS          16  
3                       DRY  NO DEFECTS          22  
4                       DRY  NO DEFECTS          14  
...                     ...         ...         ...  
433054                  DRY  NO DEFECTS          16  
433055                  DRY  NO DEFECTS          14  
433056                  DRY  NO DEFECTS          10  
433057                  DRY  NO DEFECTS          18  
433058                  DRY  NO DEFECTS           8  

[433059 rows x 9 columns]

In [38]:
transform["WEATHER_CONDITION"].value_counts(normalize=True) * 100

1     79.246246
2      9.053732
0      4.301261
3      3.624910
4      2.999360
5      0.314738
6      0.182654
7      0.151481
8      0.089826
9      0.020782
10     0.014548
11     0.000462
Name: WEATHER_CONDITION, dtype: float64

<br>

# 5. LIGHTING_CONDITION

In [39]:
def change5(x):
    if x=="DAYLIGHT": return 1
    elif x=="DARKNESS, LIGHTED ROAD": return 2
    elif x=="DARKNESS": return 3
    elif x=="DUSK": return 4
    elif x=="DAWN": return 5
    elif x=="UNKNOWN": return 0
    
transform["LIGHTING_CONDITION"] = transform["LIGHTING_CONDITION"].apply(change5)

In [40]:
transform

POSTED_SPEED_LIMIT  TRAFFIC_CONTROL_DEVICE  DEVICE_CONDITION  \
0                       10                       1                 1   
1                       45                       1                 1   
2                       35                       3                 2   
3                       30                       2                 2   
4                       35                       1                 1   
...                    ...                     ...               ...   
433054                  30                       1                 1   
433055                  30                       1                 1   
433056                  30                       2                 2   
433057                  30                       1                 1   
433058                  15                       1                 1   

        WEATHER_CONDITION  LIGHTING_CONDITION  LANE_CNT ROADWAY_SURFACE_COND  \
0                       1                   3       NaN                  DRY   
1                       1                   1       NaN                  DRY   
2                       1                   1       4.0                  DRY   
3                       1                   2       4.0                  DRY   
4                       1                   1       NaN                  DRY   
...                   ...                 ...       ...                  ...   
433054                  1                   1       NaN                  DRY   
433055                  1                   1       NaN                  DRY   
433056                  1                   1       NaN                  DRY   
433057                  1                   1       NaN                  DRY   
433058                  1                   1       NaN                  DRY   

       ROAD_DEFECT  CRASH_HOUR  
0       NO DEFECTS          18  
1       NO DEFECTS          16  
2       NO DEFECTS          16  
3       NO DEFECTS          22  
4       NO DEFECTS          14  
...            ...         ...  
433054  NO DEFECTS          16  
433055  NO DEFECTS          14  
433056  NO DEFECTS          10  
433057  NO DEFECTS          18  
433058  NO DEFECTS           8  

[433059 rows x 9 columns]

In [41]:
transform["LIGHTING_CONDITION"].value_counts(normalize=True) * 100

1    65.650177
2    20.911469
3     5.037189
0     3.603897
4     3.061246
5     1.736022
Name: LIGHTING_CONDITION, dtype: float64

<br>
# 6. LANE_CNT
**상의 ㄱㄱ** <br>
**결측치만 234102건 -> 일단 결측치 : 0으로**

<br>

# 7. ROADWAY_SURFACE_COND

In [42]:
def change7(x):
    if x=="DRY": return 1
    elif x=="WET": return 2
    elif x=="SNOW OR SLUSH": return 3
    elif x=="ICE": return 4
    elif x=="OTHER": return 5
    elif x=="SAND, MUD, DIRT": return 6
    elif x=="UNKNOWN": return 0
    
transform["ROADWAY_SURFACE_COND"] = transform["ROADWAY_SURFACE_COND"].apply(change7)

In [43]:
transform

POSTED_SPEED_LIMIT  TRAFFIC_CONTROL_DEVICE  DEVICE_CONDITION  \
0                       10                       1                 1   
1                       45                       1                 1   
2                       35                       3                 2   
3                       30                       2                 2   
4                       35                       1                 1   
...                    ...                     ...               ...   
433054                  30                       1                 1   
433055                  30                       1                 1   
433056                  30                       2                 2   
433057                  30                       1                 1   
433058                  15                       1                 1   

        WEATHER_CONDITION  LIGHTING_CONDITION  LANE_CNT  ROADWAY_SURFACE_COND  \
0                       1                   3       NaN                     1   
1                       1                   1       NaN                     1   
2                       1                   1       4.0                     1   
3                       1                   2       4.0                     1   
4                       1                   1       NaN                     1   
...                   ...                 ...       ...                   ...   
433054                  1                   1       NaN                     1   
433055                  1                   1       NaN                     1   
433056                  1                   1       NaN                     1   
433057                  1                   1       NaN                     1   
433058                  1                   1       NaN                     1   

       ROAD_DEFECT  CRASH_HOUR  
0       NO DEFECTS          18  
1       NO DEFECTS          16  
2       NO DEFECTS          16  
3       NO DEFECTS          22  
4       NO DEFECTS          14  
...            ...         ...  
433054  NO DEFECTS          16  
433055  NO DEFECTS          14  
433056  NO DEFECTS          10  
433057  NO DEFECTS          18  
433058  NO DEFECTS           8  

[433059 rows x 9 columns]

In [44]:
transform["ROADWAY_SURFACE_COND"].value_counts(normalize=True) * 100

1    75.102238
2    13.828832
0     6.555920
3     3.449876
4     0.786267
5     0.230685
6     0.046183
Name: ROADWAY_SURFACE_COND, dtype: float64

<br>

# 8. ROAD_DEFECT

In [45]:
def change8(x):
    if x=="NO DEFECTS": return 1
    elif x=="RUT, HOLES": return 2
    elif x=="OTHER": return 3
    elif x=="WORN SURFACE": return 4
    elif x=="SHOULDER DEFECT": return 5
    elif x=="DEBRIS ON ROADWAY": return 6
    elif x=="UNKNOWN": return 0
    
transform["ROAD_DEFECT"] = transform["ROAD_DEFECT"].apply(change8)

In [46]:
transform

POSTED_SPEED_LIMIT  TRAFFIC_CONTROL_DEVICE  DEVICE_CONDITION  \
0                       10                       1                 1   
1                       45                       1                 1   
2                       35                       3                 2   
3                       30                       2                 2   
4                       35                       1                 1   
...                    ...                     ...               ...   
433054                  30                       1                 1   
433055                  30                       1                 1   
433056                  30                       2                 2   
433057                  30                       1                 1   
433058                  15                       1                 1   

        WEATHER_CONDITION  LIGHTING_CONDITION  LANE_CNT  ROADWAY_SURFACE_COND  \
0                       1                   3       NaN                     1   
1                       1                   1       NaN                     1   
2                       1                   1       4.0                     1   
3                       1                   2       4.0                     1   
4                       1                   1       NaN                     1   
...                   ...                 ...       ...                   ...   
433054                  1                   1       NaN                     1   
433055                  1                   1       NaN                     1   
433056                  1                   1       NaN                     1   
433057                  1                   1       NaN                     1   
433058                  1                   1       NaN                     1   

        ROAD_DEFECT  CRASH_HOUR  
0                 1          18  
1                 1          16  
2                 1          16  
3                 1          22  
4                 1          14  
...             ...         ...  
433054            1          16  
433055            1          14  
433056            1          10  
433057            1          18  
433058            1           8  

[433059 rows x 9 columns]

In [47]:
transform["ROAD_DEFECT"].value_counts(normalize=True) * 100

1    83.333449
0    14.392034
2     0.982314
3     0.574287
4     0.408720
5     0.219831
6     0.089364
Name: ROAD_DEFECT, dtype: float64

<br>

# 9. CRASH_HOUR
**상의 ㄱㄱ**<br>
**시간대를 아침/점심/저녁으로 나눌지 말지?**